File: solutions/typescript/triangle/1/triangle.ts
export class Triangle {
  a: number;
  b: number;
  c: number;

  constructor(...sides: number[]) {
    this.a = sides[0]
    this.b = sides[1]
    this.c = sides[2]
  }

  get isEquilateral(): boolean {
    return this.isValid && (
      this.a === this.b && this.b === this.c
      )
  }

  get isIsosceles(): boolean {
    return this.isValid && (
      this.a === this.b || this.a === this.c || this.b === this.c
      )
  }

  get isScalene(): boolean {
    return this.isValid && (
      !this.isEquilateral && !this.isIsosceles
    ) 
  }

  get isValid(): boolean {
        if (this.a <= 0 || this.b <= 0 || this.c <= 0) {
            return false;
        }

        return (
            this.a + this.b > this.c &&
            this.a + this.c > this.b &&
            this.b + this.c > this.a
        );
    }
}

File: solutions/kotlin/triangle/1/src/main/kotlin/Triangle.kt
class Triangle<T : Number>(val a: T, val b: T, val c: T) {

    val x = a.toDouble()
    val y = b.toDouble()
    val z = b.toDouble()

    val sides = listOf(x, y, z).sorted()
    val sideSize = setOf(sides).size

    val isEquilateral: Boolean = isValid() && sideSize == 1
    val isIsosceles: Boolean = isValid() && sideSize <= 2
    val isScalene: Boolean = isValid() && sideSize == 3

    fun isValid(): Boolean {
        return sides.slice(0..1).sum() > sides[2]

    }
}
